mitodl/ocw-studio
# Unpublish sites - backend code

#### Pre-Flight checklist


#### What are the relevant tickets?
Partially addresses #1164 

#### What's this PR do?
Implements backend code for fulfilling the requirements of #1164:
- new database fields for `Website` (`unpublished`, `unpublished_status`, `last_unpublished_by`)
- new concourse pipeline to remove all content for unpublished sites from the publish bucket and purge the fastly cache
- new API endpoint to unpublish a site and trigger the new pipeline via a task
- new API endpoint to return a list of unpublished sites with data needed by the new pipeline.
- modified the concourse webhook endpoint to update statuses of unpublished sites

#### How should this be manually tested?
- Follow the README instructions for local concourse settings.
- Run `docker-compose --profile concourse up`
- Run `docker-compose run web python manage.py upsert_site_removal_pipeline` and verify the pipeline is created in your local concourse

Testing API endpoints:
- Go to `http://localhost:8043/api/websites/<site-name>/unpublish/` and click the POST button.  Verify that:
    - `Website.unpublished=True` and `Website.unpublished_status=not-started`
    - The `remove-unpublished-sites` pipeline starts
    - The individual site pipeline is paused (you might need to create & unpause it first).
- Run `curl -X GET -H "Authorization: Bearer <settings.API_BEARER_TOKEN>" "http://localhost:8043/api/publish/?version=live"` and verify that the site is not in the list.
- Run `curl -X GET -H "Authorization: Bearer <settings.API_BEARER_TOKEN>" "http://localhost:8043/api/unpublish/"` and verify that it is in the list.
- Run `curl -X POST  -H "Authorization: Bearer <settings.API_BEARER_TOKEN>" -H "Content-Type: application/json" "http://localhost:8043/api/websites/<site-name>/pipeline_status/" -d '{"version": "live", "status": "succeeded", "unpublished": true}' then verify that the `Website.unpublished_status` is `succeeded`
- Run `curl -X GET -H "Authorization: Bearer <settings.API_BEARER_TOKEN>" "http://localhost:8043/api/unpublish/"` again and verify that it is NOT in the list.
- Run `curl -X POST  -H "Authorization: Bearer <settings.API_BEARER_TOKEN>" -H "Content-Type: application/json" "http://localhost:8043/api/websites/<site-name>/pipeline_status/" -d '{"version": "live", "status": "errored", "unpublished": true}' then verify that the `Website.unpublished_status` is `errored`
- Run `curl -X GET -H "Authorization: Bearer <settings.API_BEARER_TOKEN>" "http://localhost:8043/api/unpublish/"` again and verify that it is now in the list again.

Testing the pipeline
- A test pipeline was created [here](https://cicd-qa.odl.mit.edu/teams/ocw/pipelines/mbtestunpub).  If you would like to try it yourself, you can do the following:
- Run `ngrok http 8043`
- In your `.env` file, set `OCW_STUDIO_BASE_URL=<your ngrok url>`
- Also set the `CONCOURSE_*` settings, plus `API_BEARER_TOKEN`, `OCW_NEXT_SEARCH_WEBHOOK_KEY`, and `OPEN_DISCUSSIONS_URL` to the same as RC
- Run `manage.py upsert_site_removal_pipeline`
- Create a site in RC with the same name as the local one you set to unpublished earlier.
- Run the pipeline, it should succeed
- Verify that the `Website.unpublished_status` is `succeeded` for your local instance site.
- Set the CONCOURSE_* settings in your .env file back to a local instance.

#### Where should the reviewer start?
Read the issue and linked RFC

#### Any background context you want to provide?
Front-end UI for unpublishing is TBD and will be handled in a separate PR.
The open-discussions webhook endpoint also [needs to be updated](https://github.com/mitodl/open-discussions/issues/3561) to handle the new `deleted=true` parameter.
